### PR TITLE
Port Glacier2 to C++11 (Fixes #629)

### LINKED ISSUE #629: Port Glacier2 to C++11
In addition to porting to C++11 we can also:
- Remove batch and associated properties
- Remove remove session thread and use ACM
- Remove `Glacier2.SesionTimeout`
- Switch to unbuffered mode by default

### PR DESCRIPTION
**Pull request by externl**

### CODE REVIEW

File: cpp/src/Glacier2/Blobject.cpp [hunk 1]
```diff
@@ -16,25 +16,20 @@ namespace
...
 const string clientTraceOverride = "Glacier2.Client.Trace.Override";
 
 }
 
-Glacier2::Blobject::Blobject(const InstancePtr& instance, const ConnectionPtr& reverseConnection,
+Glacier2::Blobject::Blobject(shared_ptr<Instance> instance, shared_ptr<Connection> reverseConnection,
                              const Context& context) :
-    _instance(instance),
+    _instance(move(instance)),
     _reverseConnection(reverseConnection),
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.cpp]: move missing

File: cpp/src/Glacier2/Blobject.cpp [hunk 2]
```diff
@@ -286,12 +209,12 @@ Glacier2::Blobject::invoke(ObjectPrx& proxy, const AMD_Object_ice_invokePtr& amd
...
-            override = _requestQueue->addRequest(new Request(proxy, inParams, current, _forwardContext, _context,
-                                                             amdCB));
+            override = _requestQueue->addRequest(make_shared<Request>(proxy, inParams, current, _forwardContext,
+                                                                      _context, move(response), move(exception)));
         }
-        catch(const ObjectNotExistException& ex)
+        catch(const ObjectNotExistException&)
         {
-            amdCB->ice_exception(ex);
+            exception(current_exception());
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.cpp]: exception was moved in the call above

File: cpp/src/Glacier2/Blobject.cpp [hunk 3]
```diff
@@ -350,28 +308,32 @@ Glacier2::Blobject::invoke(ObjectPrx& proxy, const AMD_Object_ice_invokePtr& amd
...
+                    proxy->ice_invokeAsync(current.operation, current.mode, inParams,
+                                           move(amiResponse), move(amiException), move(amiSent));
                 }
             }
         }
-        catch(const LocalException& ex)
+        catch(const LocalException&)
         {
-            amdCB->ice_exception(ex);
+            exception(current_exception());
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.cpp]: exception was moved earlier

File: cpp/src/Glacier2/Blobject.cpp [hunk 4]
```diff
@@ -327,21 +250,56 @@ Glacier2::Blobject::invoke(ObjectPrx& proxy, const AMD_Object_ice_invokePtr& amd
...
-        // If we are in not in buffered mode, we send the request
-        // directly.
+        // If we are in not in buffered mode, we send the request directly.
         //
-        assert(!proxy->ice_isBatchOneway() && !proxy->ice_isBatchDatagram());
 
         try
         {
-            Callback_Object_ice_invokePtr amiCB;
+            function<void(bool, const pair<const Byte*, const Byte*>&)> amiResponse = nullptr;
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.cpp]: the second parameter of this AMI response is a pair, not a const pair&.

File: cpp/src/Glacier2/Blobject.h [hunk 5]
```diff
@@ -12,37 +12,35 @@
...
 
-class Blobject : public Ice::BlobjectArrayAsync
+class Blobject : public Ice::BlobjectArrayAsync, public std::enable_shared_from_this<Blobject>
 {
 public:
 
-    Blobject(const InstancePtr&, const Ice::ConnectionPtr&, const Ice::Context&);
-    virtual ~Blobject();
+    Blobject(std::shared_ptr<Instance>, std::shared_ptr<Ice::Connection>, const Ice::Context&);
+    virtual ~Blobject() override;
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.h]: I would remove this default dtor

File: cpp/src/Glacier2/Blobject.h [hunk 6]
```diff
@@ -12,37 +12,35 @@
...
-    void invokeResponse(bool, const std::pair<const Ice::Byte*, const Ice::Byte*>&,
-                        const Ice::AMD_Object_ice_invokePtr&);
-    void invokeSent(bool, const Ice::AMD_Object_ice_invokePtr&);
-    void invokeException(const Ice::Exception&, const Ice::AMD_Object_ice_invokePtr&);
+    void invokeException(std::exception_ptr, std::function<void(std::exception_ptr)>&&);
 
 protected:
 
-    void invoke(Ice::ObjectPrx&, const Ice::AMD_Object_ice_invokePtr&,
+    void invoke(std::shared_ptr<Ice::ObjectPrx>&,
```
Comment (bernardnormier) [inline cpp/src/Glacier2/Blobject.h]: I would reorder the parameters to match invokeAsync's order

File: cpp/src/Glacier2/ClientBlobject.h [hunk 7]
```diff
@@ -11,35 +11,33 @@
...
-class ClientBlobject : public Glacier2::Blobject
+class ClientBlobject final : public Glacier2::Blobject
 {
 public:
 
-    ClientBlobject(const InstancePtr&, const FilterManagerPtr&, const Ice::Context&, const RoutingTablePtr&);
-    virtual ~ClientBlobject();
+    ClientBlobject(std::shared_ptr<Instance>, std::shared_ptr<FilterManager>, const Ice::Context&,
+                   std::shared_ptr<RoutingTable>);
+    ~ClientBlobject() override;
```
Comment (bernardnormier) [inline cpp/src/Glacier2/ClientBlobject.h]: Can be removed

File: cpp/src/Glacier2/ServantLocator.h [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+#define GLACIER2_LOCATOR_H
+
+#include <Ice/Ice.h>
+
+#include <Glacier2/SessionRouterI.h>
+
+namespace Glacier2
+{
+
+class ClientLocator final : public Ice::ServantLocator
```
Comment (bernardnormier) [inline cpp/src/Glacier2/ServantLocator.h]: Could move into the .cpp file that uses these servant locators.

File: cpp/src/Glacier2/SessionRouterI.cpp [hunk 9]
```diff
@@ -1015,49 +870,49 @@ SessionRouterI::getACMTimeout(const Ice::Current& current) const
 void
 SessionRouterI::updateSessionObservers()
 {
-    IceUtil::Monitor<IceUtil::Mutex>::Lock lock(*this);
+    lock_guard<mutex> lg(_mutex);
 
-    Glacier2::Instrumentation::RouterObserverPtr observer = _instance->getObserver();
+    auto& observer = _instance->getObserver();
```
Comment (bernardnormier) [inline cpp/src/Glacier2/SessionRouterI.cpp]: Looks like it should be auto

Comment (bernardnormier) [inline cpp/src/Glacier2/SessionRouterI.cpp]: Should this be an auto const&?

File: cpp/src/Glacier2/RequestQueue.h [hunk 10]
```diff
@@ -17,99 +15,101 @@ namespace Glacier2
...
-
-    Ice::AsyncResultPtr invoke(const Ice::Callback_Object_ice_invokePtr& callback);
-    bool override(const RequestPtr&) const;
-    void addBatchProxy(std::set<Ice::ObjectPrx>&);
+    Request(std::shared_ptr<Ice::ObjectPrx>,
+            const std::pair<const Ice::Byte*, const Ice::Byte*>&,
+            const Ice::Current&,
+            bool,
+            const Ice::Context&,
+            std::function<void(bool, const std::pair<const Ice::Byte*, const Ice::Byte*>&)>&&,
```
Comment (bernardnormier) [inline cpp/src/Glacier2/RequestQueue.h]: Should the pair&& be non-const?

Comment (externl) [reply cpp/src/Glacier2/RequestQueue.h]: Yes, I think so.

File: cpp/src/Glacier2CryptPermissionsVerifier/CryptPermissionsVerifierI.cpp [hunk 11]
```diff
@@ -32,6 +32,550 @@ using namespace Ice;
...
+namespace
+{
+
+#if defined(__FreeBSD__) && !defined(__GLIBC__)
+
+//
+// FreeBSD crypt is no reentrat we use this global mutex
+// to serialize access.
+//
+IceUtil::Mutex* _staticMutex = 0;
```
Comment (bernardnormier) [inline cpp/src/Glacier2CryptPermissionsVerifier/CryptPermissionsVerifierI.cpp]: Should use a plain global std::mutex!

File: cpp/src/Glacier2CryptPermissionsVerifier/CryptPermissionsVerifierI.cpp [hunk 12]
```diff
@@ -32,6 +32,550 @@ using namespace Ice;
...
+};
+
+Init init;
+
+#elif defined(__APPLE__)
+
+// UniqueRef helper class for CoreFoundation classes, comparable to std::unique_ptr
+
+template<typename R>
+class UniqueRef
```
Comment (bernardnormier) [inline cpp/src/Glacier2CryptPermissionsVerifier/CryptPermissionsVerifierI.cpp]: should use a unique_ptr with a custom Deleter instead

File: cpp/src/Glacier2/RequestQueue.cpp [hunk 13]
```diff
@@ -518,24 +428,25 @@ Glacier2::RequestQueueThread::run()
...
                 if(_sleep)
                 {
-                    IceUtil::Time now = IceUtil::Time::now(IceUtil::Time::Monotonic);
-                    if(!timedWait(_sleepDuration))
+                    auto now = chrono::steady_clock::now();
+
+                    if(_condVar.wait_for(lock, _sleepDuration) == cv_status::no_timeout)
                     {
-                        _sleepDuration = IceUtil::Time();
+                        _sleepDuration = 0ns;
```
Comment (bernardnormier) [inline cpp/src/Glacier2/RequestQueue.cpp]: _sleepDuration, used only in run, should not be a data member

File: cpp/src/Glacier2/SessionRouterI.cpp [hunk 14]
```diff
@@ -320,136 +220,108 @@ class SSLCreateSession : public CreateSession
...
-        _sessionRouter->_sslVerifier->begin_authorize(_sslInfo, ctx,
-                                                      newCallback_SSLPermissionsVerifier_authorize(this,
-                                                                             &SSLCreateSession::authorizeResponse,
-                                                                             &SSLCreateSession::authorizeException));
+
+        auto self = static_pointer_cast<SSLCreateSession>(shared_from_this());
+        _sessionRouter->_sslVerifier->authorizeAsync(_sslInfo,
+                                                     [self](bool ok, const string& reason)
+                                                     {
+                                                         self->authorizeResponse(ok, reason);
```
Comment (bernardnormier) [inline cpp/src/Glacier2/SessionRouterI.cpp]: I would get rid of the helper functions called only once from AMI callbacks:
authorizeResponse|Exception,  checkPermissionsResponse|Exception

**Pull request merged**